### ISSUE TITLE
python_platlib and python_purelib break when using inheritance with packages.

### ISSUE THREAD

**Issue by Paul-Ferrell** [LLNL/spack#31648]
### Steps to reproduce

- Create a separate package repo, and add that repo to Spack.
- Pick a package that depends on python and uses `python_platlib` or `python_purelib` in its `install()` method.
- Create a new package of the same name in your separate repo. 
  - Import the needed spack libs
  - Import the parent package of the same name from builtins
  - Create a new class for that package that inherits from the parent. 
- When installing that package, spack should fail because `python_platlib` isn't defined.

Example package def (LLVM): 
```python
from spack import *
from spack.pkg.builtin.llvm import Llvm as BuiltinLlvm

class Llvm(BuiltinLlvm):
   depends_on("swig@3:", when="+lldb")  
```

This happens because the monkey patch that inserts `python_platlib` (etc) from the python package inserts it into the module
with the child class, but the `install()` method is defined in the scope of the parent class and will never see those variables.

### Error message

```console
==> Error: NameError: name 'python_platlib' is not defined

/tmp/pflarr/2022-2/linux-rhel7-x86_64_v3-omnipath-none/spack/var/spack/repos/builtin/packages/llvm/package.py:743, in post_install:
        740                ninja()
        741                ninja("install")
        742        if "+python" in self.spec:
  >>    743            install_tree("llvm/bindings/python", python_platlib)
        744
        745            if "+clang" in self.spec:
        746                install_tree("clang/bindings/python", python_platlib)
```

### Information on your system

* **Spack:** 0.19.0.dev0 (c4efeab9763b88bf40287ea77d7791b6f340b2e3)
* **Python:** 3.6.8
* **Platform:** linux-rhel7-broadwell
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by adamjstewart** [comment]
@alalazo any idea how to solve this?

**Comment by Paul-Ferrell** [comment]
You could walk the inheritance tree and set it for each module you encounter. You'd want to stop when you inherit from one of the build classes, or when the module isn't named `package.py`.

Or maybe stop monkey patching, and just provide a base methods/properties that look at the spec dependencies tree, finds Python and grabs and returns those values.

**Comment by iarspider** [comment]
@alalazo gentle ping.

**Comment by wdconinc** [comment]
I am apparently also affected by this :-\

**Comment by wdconinc** [comment]
https://github.com/spack/spack/issues/32336 indicates this is solved in https://github.com/spack/spack/pull/32340. I'm on v0.18.1 so I'll have to backport that.

**Comment by alalazo** [comment]
Should be solved by #32340 and #34059
